NEWS.md

```diff
@@ -102,6 +102,16 @@ Note that each entry is kept to a minimum, see links for details.
 
 Note: We're only listing outstanding class updates.
 
+* IO
+    * Introduce `IO#timeout=` and `IO#timeout` which can cause
+    `IO::TimeoutError` to be raised if a blocking operation exceeds the
+    specified timeout. [[Feature #18630]]
+
+    ```ruby
+    STDIN.timeout = 1
```

Comment:
Sorry for asking here, but I'm reluctant to post a bug on the Ruby Redmine. Should this feature work on all `IO`, like `File` and `FD.fd_for(some_readiness_fd)`? Because for me it doesn't.

```ruby
require 'io/wait'
require 'socket'
require 'async'

begin
  STDIN.timeout = 1
  STDIN.read
rescue IO::TimeoutError
  puts "IO#timeout= worked on STDIN"
end


begin
  r, w = UNIXSocket.pair
  r.timeout = 1
  r.read
rescue IO::TimeoutError
  puts "IO#timeout= worked on UNIXSocket"
else
  puts "IO#timeout= did NOT work on UNIXSocket"
end


begin
  file = File.open '/dev/null'
  file.timeout = 1
  file.read
rescue IO::TimeoutError
  puts "IO#timeout= worked on File"
else
  puts "IO#timeout= did NOT work on File"
end


Async do
  file.read
rescue IO::TimeoutError
  puts "IO#timeout= worked on File in Async block"
else
  puts "IO#timeout= did NOT work on File in Async block"
end
```

Gives me:
```
IO#timeout= worked on STDIN
IO#timeout= worked on UNIXSocket
IO#timeout= did NOT work on File
IO#timeout= did NOT work on File in Async block
```

Reply:
I suppose timeouts should be considered advisory and/or best effort. Like, a safety net.

However, not working for File is probably to be expected, but I wonder if we can improve it. For sure, when running within Async, on the URing backend, it should be possible to make it work.